distribution/sc-proxy-dex/src/wrapped_lp_token_merge.rs

```diff
@@ -0,0 +1,189 @@
+        self.burn_deposit_tokens(caller, &deposit);
+
+        self.nft_create_tokens(&wrapped_lp_token_id, &amount, &attrs);
+        let new_nonce = self.increase_wrapped_lp_token_nonce();
+
+        self.send_nft_tokens(
+            &wrapped_lp_token_id,
+            new_nonce,
+            &amount,
+            caller,
+            &opt_accept_funds_func,
+        );
+
+        Ok(())
+    }
+    fn get_wrapped_lp_tokens_from_deposit(
+        &self,
+        deposit: &[GenericTokenAmountPair<Self::BigUint>],
+    ) -> SCResult<Vec<WrappedLpToken<Self::BigUint>>> {
+        let mut result = Vec::new();
+
+        for elem in deposit.iter() {
+            result.push(WrappedLpToken {
+                token_amount: elem.clone(),
+                attributes: self
+                    .get_wrapped_lp_token_attributes(&elem.token_id, elem.token_nonce)?,
+            })
+        }
+        Ok(result)
+    }
+
+    fn require_wrapped_lp_tokens_from_same_pair(
+        &self,
+        tokens: &[WrappedLpToken<Self::BigUint>],
+    ) -> SCResult<()> {
+        let lp_token_id = tokens[0].attributes.lp_token_id.clone();
+
+        for elem in tokens.iter() {
+            require!(
+                elem.attributes.lp_token_id == lp_token_id,
+                "Lp token id differs"
+            );
+        }
+        Ok(())
+    }
+
+    fn require_all_tokens_are_wrapped_lp_tokens(
+        &self,
+        tokens: &[GenericTokenAmountPair<Self::BigUint>],
+        wrapped_lp_token_id: &TokenIdentifier,
+    ) -> SCResult<()> {
+        for elem in tokens.iter() {
+            require!(
+                &elem.token_id == wrapped_lp_token_id,
+                "Not a Wrapped Lp Token"
+            );
+        }
+        Ok(())
+    }
+
+    fn get_merged_wrapped_lp_token_attributes(
+        &self,
+        tokens: &[WrappedLpToken<Self::BigUint>],
+        merged_locked_asset_token_amount: &GenericTokenAmountPair<Self::BigUint>,
+    ) -> WrappedLpTokenAttributes<Self::BigUint> {
+        let mut lp_token_amount = 0u64.into();
+
+        tokens
+            .iter()
+            .for_each(|x| lp_token_amount += &x.attributes.lp_token_total_amount);
+        WrappedLpTokenAttributes {
+            lp_token_id: tokens[0].attributes.lp_token_id.clone(),
+            lp_token_total_amount: lp_token_amount,
+            locked_assets_invested: merged_locked_asset_token_amount.amount.clone(),
+            locked_assets_nonce: merged_locked_asset_token_amount.token_nonce,
+        }
+    }
+
+    fn merge_locked_asset_tokens_from_wrapped_lp(
+        &self,
+        tokens: &[WrappedLpToken<Self::BigUint>],
+    ) -> GenericTokenAmountPair<Self::BigUint> {
+        let locked_asset_factory_addr = self.locked_asset_factory_address().get();
+        let locked_asset_token = self.locked_asset_token_id().get();
+
+        if tokens.len() == 1 {
+            let token = tokens[0].clone();
+
+            let amount = self.rule_of_three(
```

Comment:
I'm trying to figure out how this `rule_of_three` call works. As far as I can see (though I probably made a mistake somewhere, the code is a bit complex), this takes a fraction of `token.token_amount.amount/token.attributes.lp_token_total_amount`, which can be up to `2`, out of `token.attributes.locked_assets_invested`. Why does it make sense for the amount to be twice the value of `token.attributes.locked_assets_invested`?

Let's take an example:

1. The user adds `asset_amount` of `locked_asset_token_id()` as liquidity to a pair contract. This gets recorded in a `WrappedLpTokenAttributes` struct (with `lp_token_total_amount=lp_amount` and `token.attributes.locked_assets_invested=asset_amount`), and the user receives `lp_amount` of `wrapped_farm_token_id()`. Note that the relation between `asset_amount` and `lp_amount` is not easy to describe (they are not necessarily equal).
2. The user adds `lp_amount` of `wrapped_farm_token_id()` to a farm contract, with locked rewards. This gets recorded in a `WrappedFarmTokenAttributes` structure, and the user receives `2*lp_amount` of `wrapped_farm_token_id()`.
3. The user takes this `2*lp_amount` of `wrapped_farm_token_id()` and deposits them. In principle, the user may deposit other amounts, also, in which case we will reach the similar `rule_of_three` line below.
4. The user calls `merge_wrapped_farm_tokens`. We can reach this line with the following sequence of calls: `merge_wrapped_farm_tokens -> merge_wrapped_farm_tokens_and_send -> merge_farming_tokens -> merge_wrapped_lp_tokens_from_farm -> merge_locked_asset_tokens_from_wrapped_lp`
5. This line calls rule_of_three(2 * lp_amount, lp_amount, asset_amount), with a result of `2*asset_amount`.

Did I make any mistakes? If not, why does it make sense to have `amount = 2 * asset_amount`? I think that this gets recorded again as a `locked_assets_invested` field, on which the user can probably apply steps 3-5 again.

Reply:
The 2 * asset_amount breaks when using Locked Mex to farm at MEX Farm after Compunding Rewards. After doing this, the relation between locked assets invested and farm tokens received breaks, meaning wrapped farm tokens amount won't be equal with 2 * locked assets invested anymore.

Reply:
+ the apr multiplier can change later. The requirement was such that it can be changed for the new positions at any time. And  Proxy SC having to rely on a setting inside the farm token attributes (so a token it cannot create or burn) doesn't seem right

Reply:
Thanks for explaining on Google Meet, my mistake was in steps 4-5: `merge_farming_tokens` would convert `2*lp_amount` to `lp_amount`, so the call in step 5 would use `lp_amount` instead of `2*lp_amount`.

Resolved.